sql/plan/delete.go

```diff
@@ -136,60 +321,88 @@ func (d *deleteIter) Next(ctx *sql.Context) (sql.Row, error) {
 	default:
 	}
 
-	// Reduce the row to the length of the schema. The length can differ when some update values come from an outer
-	// scope, which will be the first N values in the row.
-	// TODO: handle this in the analyzer instead?
-	if len(d.schema) < len(row) {
-		row = row[len(row)-len(d.schema):]
+	// For each target table from which we are deleting rows, reduce the row from our child iterator to just
+	// the columns that are part of that target table. This means looking at the position in the schema for
+	// the target table and also removing any prepended columns contributed by outer scopes.
+	fullSchemaLength := len(d.schema)
+	rowLength := len(row)
+	for _, deleter := range d.deleters {
+		schemaLength := deleter.schemaEnd - deleter.schemaStart
+		subSlice := row
+		if schemaLength < rowLength {
+			subSlice = row[(rowLength - fullSchemaLength + deleter.schemaStart):(rowLength - fullSchemaLength + deleter.schemaEnd)]
+		}
+		err = deleter.deleter.Delete(ctx, subSlice)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	return row, d.deleter.Delete(ctx, row)
+	return row, nil
 }
 
 func (d *deleteIter) Close(ctx *sql.Context) error {
 	if !d.closed {
 		d.closed = true
-		if err := d.deleter.Close(ctx); err != nil {
-			return err
+		for _, deleter := range d.deleters {
+			if err := deleter.deleter.Close(ctx); err != nil {
```

Comment:
i think we might still need to close all of them if one errors?

Reply:
Changed this to ensure we close all deleters and the childIter. We now track the first error encountered and return it after closing all resources.